coreax/solvers/recombination.py

```diff
@@ -0,0 +1,570 @@
+    :param test_functions: A callable that applies a set of specified test-functions
+        :math:`\Phi^\prime = \{\phi_1,\dots,\phi_{m-1}\}` where each function is a map
+        :math:`\phi_i \colon \Omega \to \mathbb{R}`; a value of non implies the identity
+        map :math:`\Phi^\prime \colon x \mapsto x`, and necessarily assumes that
+        :math:`x \in \Omega \subseteq \mathbb{R}^{m-1}`
+    :param augment: If to prepend prepend the affine-augmentation test function
+        :math:`\{x \mapsto 1\}` to the explicitly pushed forward nodes \Phi^\prime(x),
+        to yield \Phi(x)
+    :return: The pushed-forward nodes.
+    """
+    if test_functions is None:
+        push_forward_nodes = nodes
+    else:
+        push_forward_nodes = jax.vmap(test_functions, in_axes=0)(nodes)
+    if augment:
+        shape, dtype = push_forward_nodes.shape[0], push_forward_nodes.dtype
+        affine_augmentation = jnp.ones((shape,), dtype)
+        push_forward_nodes = jnp.c_[affine_augmentation, push_forward_nodes]
+    return push_forward_nodes
+
+
+def _co_linearize(
+    nodes: Shaped[Array, "n m"], weights: Shaped[Array, " n"]
+) -> tuple[Shaped[Array, "n m"], Shaped[Array, " n"], Shaped[Array, " n"]]:
+    """
+    Make zero-weighted nodes co-linear with the maximum weighted node.
+
+    Due to the static shape requirements imposed by JAX, we implicitly remove nodes by
+    setting their corresponding weight to zero. This is sufficient in the recombination
+    algorithm for all but one scenario, the computation of the null space basis. Because
+    the zero-weighted nodes still exist in the node matrix, they influence the SVD and
+    yield an erroneous null space basis.
+
+    We ameliorate this problem by setting the zero-weighted nodes equal (co-linear) to
+    the largest weighted node (an arbitrary but consistent choice). Because the nodes
+    are now co-linear to each other and the largest weighted node, we know that at least
+    all but one of them can be safely eliminated by the recombination procedure. Thus,
+    the nodes become effectively "invisible" to the elimination procedure.
+
+    The only caveat is that we don't know which of the equal nodes will be retained post
+    elimination. To handle this, we keep an index (reference) from the zero-weighted
+    nodes to the largest weighted node, and we redistribute the largest weight equally
+    over all the "co-linearized" nodes (preserving the CoM and allowing any node to be
+    eliminated).
+
+    :param nodes: The nodes to co-linearize
+    :param weights: The weights to apply the co-linearization correction to
+    :return: The co-linearized nodes, corrected weights, and co-linearized-to-original
+        reference indices.
+    """
+    max_index = jnp.argmax(weights)
+    non_zero_weights_mask = weights > 0
+    zero_weights_mask = 1 - non_zero_weights_mask
+    n_zeros = zero_weights_mask.sum()
+    weights = weights.at[max_index].divide(n_zeros + 1)
```

Comment:
Some extra comments describing what each step does here would be helpful for me!

Reply:
Really helpful comments, thanks!